samza-core/src/main/java/org/apache/samza/task/StreamOperatorTask.java

```diff
@@ -18,113 +18,114 @@
+   * and the logical transforms using the {@link org.apache.samza.operators.MessageStream} APIs.
+   *<p>
+   * It then uses the {@link StreamGraphImpl} to create the {@link OperatorImplGraph} corresponding to the logical
+   * DAG. It also saves the mapping between input {@link SystemStream}s and their corresponding {@link InputStream}s
+   * for delivering incoming messages to the appropriate sub-DAG.
+   *
+   * @param config allows accessing of fields in the configuration files that this StreamTask is specified in
+   * @param context allows initializing and accessing contextual data of this StreamTask
+   * @throws Exception in case of initialization errors
+   */
   @Override
   public final void init(Config config, TaskContext context) throws Exception {
-    // create the MessageStreamsImpl object and initialize app-specific logic DAG within the task
-    StreamGraphImpl streamGraph = new StreamGraphImpl(this.runner, config);
-    this.graphBuilder.init(streamGraph, config);
-    // get the context manager of the {@link StreamGraph} and initialize the task-specific context
-    this.contextManager = streamGraph.getContextManager();
-    this.systemStreamPartitions = context.getSystemStreamPartitions();
+    StreamGraphImpl streamGraph = new StreamGraphImpl(runner, config);
+    // initialize the user-implemented stream application.
+    this.streamApplication.init(streamGraph, config);
 
-    Map<SystemStream, MessageStreamImpl> inputBySystemStream = new HashMap<>();
-    systemStreamPartitions.forEach(ssp -> {
-        if (!inputBySystemStream.containsKey(ssp.getSystemStream())) {
-          // create mapping from the physical input {@link SystemStream} to the logic {@link MessageStream}
-          inputBySystemStream.putIfAbsent(ssp.getSystemStream(), streamGraph.getInputStream(ssp.getSystemStream()));
-        }
+    // get the user-implemented context manager and initialize the task-specific context.
+    this.contextManager = streamGraph.getContextManager();
+    TaskContext initializedTaskContext = this.contextManager.initTaskContext(config, context);
+
+    // create the operator impl DAG corresponding to the logical operator spec DAG
+    OperatorImplGraph operatorImplGraph = new OperatorImplGraph(clock);
+    operatorImplGraph.init(streamGraph, config, initializedTaskContext);
+    this.operatorImplGraph = operatorImplGraph;
+
+    // TODO: SAMZA-1118 - Remove mapping after SystemConsumer starts returning logical streamId with incoming messages
+    inputSystemStreamToInputStream = new HashMap<>();
+    streamGraph.getInputStreams().forEach((streamSpec, inputStream)-> {
+        SystemStream systemStream = new SystemStream(streamSpec.getSystemName(), streamSpec.getPhysicalName());
+        inputSystemStreamToInputStream.put(systemStream, inputStream);
       });
-    operatorGraph.init(inputBySystemStream, config, this.contextManager.initTaskContext(config, context));
```

Comment:
What do you have in mind? Would you prefer passing in only inputStreams directly? I like passing in StreamGraphImpl since it makes sense as the interface - pass in the operatorSpec DAG and get back the operatorImpl DAG.